src/UseCase/BannerSelection/BannerSelection.php

```diff
@@ -9,25 +9,29 @@
  */
 class BannerSelection {
 
-	private $bannerValues = [];
-	private $displayBanner;
+	private $bannerIdentifier;
+	private $visitorData;
+	private $campaignEnd;
+
+	public function __construct( ?string $bannerIdentifier, Visitor $visitorData, \DateTime $campaignEnd ) {
+		$this->bannerIdentifier = $bannerIdentifier;
+		$this->visitorData = $visitorData;
+		$this->campaignEnd = $campaignEnd;
+	}
 
-	public function __construct( Visitor $bannerValues, bool $displayBanner ) {
-		$this->displayBanner = $displayBanner;
-		$this->bannerValues = $bannerValues;
+	public function displayBanner(): bool {
+		return $this->bannerIdentifier !== null;
 	}
 
-	public function getBannerValues(): array {
-		return $this->bannerValues;
+	public function getBannerIdentifier(): ?string {
```

Comment:
Can you elaborate on this a bit more or provide some reading material on this? I have never seen an assert used like that.

Reply:
You see that the function signature guarantees that it returns a string? The calling code of `BannerSelection` should not need to care about how the different "versions" of empty vs non-empty `BannerSelection` are implemented and should not do low-level null checks instead of using descriptive method names. When `displayBanner` returns false, `getBannerIdentifier` should never be called. But when `displayBanner` returns is true, why should `getBannerIdentifier` return a null value, we should already know by then that it will return something.

Without the assertion, phpstan would complain about the type mismatch between property and method return type. Having a guard clause in front of the return value tells PHPStan that we know what we're doing. The following code would also do the job, and be more explicit, but adds 2 more lines to a simple getter:

public function getBannerIdentifier(): string {
    if( $this->bannerIdentifier === null ) {
        throw \RuntimeException('getBannerIdentifier should not be called when no banner can be displayed'):
    }
    return $this->bannerIdentifier;
}

Using `assert` (instead of `if` or assertion libraries) has its benefits and drawbacks, I'm preparing a blog article on that.

A cleaner but more code-intensive solution would be making `BannerSelection` an abstract base class with subclasses for Empty and Non-Empty results and custom implementations of `displayBanner` and `getBannerIdentifier`. In the empty selection, `getBannerIdentifier` should throw an exception.

Pinging @tzhelyazkova because this is might also be interesting for her.

Reply:
Thanks!